File: frontend/src/api/decisionMirror.ts
import type {
  DecisionMirrorCallResponse,
  DecisionMirrorContext,
  DecisionMirrorMessagesResponse,
  DecisionMirrorProfile,
  DecisionMirrorProfileResponse,
  DecisionMirrorMessagesRequestPayload,
  DecisionMirrorScoreResponse,
  DecisionMirrorTranscriptTurn,
  Difficulty,
} from '@/types/decisionMirror';

async function parseError(response: Response): Promise<string> {
  try {
    const payload = await response.json();
    return payload?.detail || payload?.message || `${response.status} ${response.statusText}`;
  } catch {
    return `${response.status} ${response.statusText}`;
  }
}

async function requestJson<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, {
    credentials: 'include',
    ...init,
  });
  if (!response.ok) {
    throw new Error(await parseError(response));
  }
  return (await response.json()) as T;
}

export async function createDecisionMirrorProfile(
  context: DecisionMirrorContext,
): Promise<DecisionMirrorProfileResponse> {
  return requestJson<DecisionMirrorProfileResponse>('/api/decision-mirror/profile', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(context),
  });
}

export async function createDecisionMirrorMessages(payload: DecisionMirrorMessagesRequestPayload): Promise<DecisionMirrorMessagesResponse> {
  return requestJson<DecisionMirrorMessagesResponse>('/api/decision-mirror/messages', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export async function scoreDecisionMirrorMessage(payload: {
  profile: DecisionMirrorProfile;
  message: string;
  goal: string;
  constraints?: string;
}): Promise<DecisionMirrorScoreResponse> {
  return requestJson<DecisionMirrorScoreResponse>('/api/decision-mirror/score', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

export async function nextDecisionMirrorCallTurn(payload: {
  profile: DecisionMirrorProfile;
  call_goal: string;
  my_key_points: string;
  difficulty: Difficulty;
  transcript: DecisionMirrorTranscriptTurn[];
}): Promise<DecisionMirrorCallResponse> {
  return requestJson<DecisionMirrorCallResponse>('/api/decision-mirror/call', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  });
}

File: backend/app/routers/decision_mirror.py
from __future__ import annotations

from fastapi import APIRouter, Depends

from backend.app.schemas.decision_mirror import (
    DecisionMirrorCallRequest,
    DecisionMirrorCallResponse,
    DecisionMirrorMessagesRequest,
    DecisionMirrorMessagesResponse,
    DecisionMirrorProfileRequest,
    DecisionMirrorProfileResponse,
    DecisionMirrorScoreRequest,
    DecisionMirrorScoreResponse,
)
from backend.app.services.auth_helpers import get_current_user
from backend.app.services.decision_mirror_engine import DecisionMirrorEngine
from backend.models.user import User


router = APIRouter(tags=["decision-mirror"])
engine = DecisionMirrorEngine()


@router.post("/api/decision-mirror/profile", response_model=DecisionMirrorProfileResponse)
def post_decision_mirror_profile(
    body: DecisionMirrorProfileRequest,
    current_user: User = Depends(get_current_user),
):
    _ = current_user
    return engine.build_profile(context=body)


@router.post("/api/decision-mirror/messages", response_model=DecisionMirrorMessagesResponse)
def post_decision_mirror_messages(
    body: DecisionMirrorMessagesRequest,
    current_user: User = Depends(get_current_user),
):
    _ = current_user
    return engine.build_messages(
        context=body.context,
        goal=body.goal,
        constraints=body.constraints,
        question_attachments_text=body.question_attachments_text,
    )


@router.post("/api/decision-mirror/score", response_model=DecisionMirrorScoreResponse)
def post_decision_mirror_score(
    body: DecisionMirrorScoreRequest,
    current_user: User = Depends(get_current_user),
):
    _ = current_user
    return engine.score_message(
        profile=body.profile,
        message=body.message,
        goal=body.goal,
        constraints=body.constraints,
    )


@router.post("/api/decision-mirror/call", response_model=DecisionMirrorCallResponse)
def post_decision_mirror_call(
    body: DecisionMirrorCallRequest,
    current_user: User = Depends(get_current_user),
):
    _ = current_user
    transcript = [item.model_dump() for item in body.transcript]
    return engine.call_next(
        profile=body.profile,
        call_goal=body.call_goal,
        my_key_points=body.my_key_points,
        difficulty=body.difficulty,
        transcript=transcript,
    )


